### ISSUE TITLE
Uncaught panic in ..applier.CreatePlanForTarget

### ISSUE THREAD

**Issue by rkorebrits-paystack** [skeema/skeema#226]
**Describe the bug**
After adding a column to the file system schema, when I run `skeema diff production [--debug]`, to get a DDL script I get this `ERROR`

**Environment**
- Skeema version and OS: 
  - `skeema version 1.11.1-community`
  - `commit d637ed77fa1fe1f82f8fdd982740b99f754cf2d5, released 2023-12-04T22:50:03Z`
  - Running on MacOS
- Database vendor, version, OS/platform: MySQL 5.7 on RDS
- Field collation: `utf8mb4_0900_ai_ci`

**To Reproduce**
 - Open a `.sql` file, add a new column
   - Column added: `demo varchar(30) COLLATE utf8mb4_0900_ai_ci DEFAULT NULL,`
 - Run `skeema diff production`

**Expected behavior**
DDL Script for the update

**Additional context**
```bash
2024-03-07 16:13:34 [DEBUG] goroutine 1393 [running]:
                            runtime/debug.Stack()
                                /opt/hostedtoolcache/go/1.21.4/x64/src/runtime/debug/stack.go:24 +0x64
                            main.panicHandler()
                                /home/runner/work/skeema/skeema/exit.go:150 +0x114
                            panic({0x1051e0a60?, 0x1053d25d0?})
                                /opt/hostedtoolcache/go/1.21.4/x64/src/runtime/panic.go:914 +0x218
                            github.com/skeema/skeema/internal/applier.CreatePlanForTarget(0x1400069ec00,
                            0x140003dee10?, {0x1, 0x2, 0x0, {0x0, 0x0}, {0x0, 0x0}, 0x0, ...})
                                /home/runner/work/skeema/skeema/internal/applier/applier.go:318 +0x7d0
                            github.com/skeema/skeema/internal/applier.ApplyTarget(0x1400069ec00, {0x105231fb8,
                            0x14000011240})
                                /home/runner/work/skeema/skeema/internal/applier/applier.go:164 +0x4bc
                            main.PushHandler.func1()
                                /home/runner/work/skeema/skeema/cmd_push.go:121 +0xe8
                            golang.org/x/sync/errgroup.(*Group).Go.func1()
                                /home/runner/work/skeema/skeema/vendor/golang.org/x/sync/errgroup/errgroup.go:75 +0x58
                            created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 1
                                /home/runner/work/skeema/skeema/vendor/golang.org/x/sync/errgroup/errgroup.go:72 +0x98
                            
2024-03-07 16:13:34 [ERROR] Uncaught panic in github.com/skeema/skeema/internal/applier.CreatePlanForTarget at
                            /home/runner/work/skeema/skeema/internal/applier/applier.go:318: runtime error: invalid
                            memory address or nil pointer dereference
                            This situation indicates a bug in Skeema. Use --debug to view full stack trace.
                            Please file an issue report at https://github.com/skeema/skeema/issues with any available
                            background information.
```

**Comment by evanelias** [comment]
Thank you for the report. Some follow-up questions to help reproduce this please:

1. Are you 100% sure the server is still running MySQL 5.7? Can you share the output of `select version();` please? (The reason I ask is that the utf8mb4_0900_ai_ci collation does not exist in stock MySQL 5.7, and attempting to use it there will cause a non-panic error much earlier in the diff process, during the workspace step.)
2. Were any warnings displayed earlier in the diff output?
3. Are you certain there are no other changes present in the diff? Prior to adding the new column to the filesystem schema, is `skeema diff` blank? (The reason I ask is that the stack trace lines suggest the generated diff is being flagged as unsafe, but simply adding a new column is never an unsafe action.)
4. What is the table's default collation?
5. If you use a different collation for the new column, does the panic still happen?
6. If you're able to share the table structure, that would also help tremendously -- both the "before" version (`SHOW CREATE TABLE` from the live database) and desired "after" version (the filesystem .sql file with your changes in place). You can redact names of the table/columns/etc if needed. Totally understand if your company policies prohibit sharing this at all though.

**Comment by evanelias** [comment]
> Warnings - nothing specific, just general warnings
> `Substituting mysql:8.0 instead for workspace purposes, which may cause behavior differences.`

Thanks, these warnings are extremely important in this case and would have helped speed up the investigation tremendously, they're pointing to one of the triggers for this issue (your target db is 5.7 but your workspace is using 8.0 because of an ARM CPU). I'll update our issue template to mention to include any warnings.

> Plain `skeema diff` also panics, prob for the same reason.

Yeah, the panic has nothing to do with the contents of your diff. If you meet the exact 4 criteria in the bullet points in my previous comment, the panic will occur.

> skeema version 1.11.1-community
> MySQL 5.7 on RDS
> `select version();` => `5.7.12-log`
> `ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci`

It appears you are running MySQL on Amazon Aurora, not standard Amazon RDS. This collation does not exist in standard MySQL 5.7, it is normally only present in MySQL 8+. You only have it because Amazon [backported it into their fork of 5.7 for Aurora](https://www.percona.com/blog/differing-mysql-client-versions-causing-broken-replication-and-collations-in-aurora/), presumably to aid in customers' upgrades to MySQL 8.

In general please note that Skeema Community Edition [does not guarantee full compatibility with Aurora](https://www.skeema.io/docs/requirements/) since there is no free way to spin up an Aurora test database in CI. Only our paid products are tested against Aurora.

In this case the bug can come up outside of Aurora also, and will be fixed soon in both editions. But that isn't guaranteed to be the case in general with other bugs, since Aurora is a proprietary fork of MySQL and [does not always have the same behavior](https://www.skeema.io/docs/requirements/#aws-aurora).

**Comment by evanelias** [comment]
A fix for the panic has been committed to the main branch. Thanks again for the report.

As it may be 1-2 months until the next Skeema binary release, here are a bunch of possible workarounds in the meantime, with various trade-offs:

a) Compile Skeema from source using the latest commit: `go install github.com/skeema/skeema@main`. This requires Golang 1.21+.

b) Remove your local machine's mysql:8.0 Docker image; fetch the mysql:8.0.28 Docker image instead; and tag it as mysql:8.0. This way, Skeema will use 8.0.28 for workspace interactions. This version of MySQL is old enough to refer to utf8mb3 as simply "utf8", which avoids the problem.

c) Switch to workspace=temp-schema instead of workspace=docker. This will run workspaces on the real target database. However, that may be disruptive to production workloads if you have a lot of tables and a large buffer pool, so proceed with extreme caution.

d) Switch to workspace=temp-schema combined with temp-schema-environment premium option, to run workspaces on a real Aurora 5.7 cluster, just not your production one. This isn't available in Skeema Community Edition.

e) Run `skeema` from an Intel-based machine. This way, workspace=docker will use a real MySQL 5.7 image instead of an 8.0 image. However, this will break on your use of utf8mb4_0900_ai_ci since that doesn't exist in stock MySQL 5.7.

f) Use `skeema diff --allow-unsafe` to bypass the affected code path. This disables Skeema's unsafe-change detection though, so not recommended if you ever use `skeema push`, but maybe fine if you only use `skeema diff`.

g) Downgrade to Skeema v1.10.0, which predates the refactor that leads to the panic condition. However, this is not recommended, as you will lose some features and several releases worth of bugfixes.

**Comment by rkorebrits-paystack** [comment]
Thanks, I'm quite far removed from the actual DB and don't have a lot of insights on the DevOps side of things. 

I have been able to run the command now after `a) Compile Skeema from source`. The diff however is picking up a lot more than I changed;

1. `skeema pull production` (no changes, already up to date)
2. Add new column (single line)
3. `skeema diff production`
  *  Getting 8 `ALTER`s, instead of 1.
  * Example
    * In the `.sql` file we have `) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;`
    * When I get `TABLE_COLLATION` from `information_schema.TABLES` I see `utf8mb4_general_ci`

SQL file
```sql
CREATE TABLE `[mytable]` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `[field1]` char(3) NOT NULL,
  `[field2]` char(3) NOT NULL,
  `updated_at` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP,
  `created_at` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
```

Generated ALTER:
```sql
ALTER TABLE `[mytable]` 
  DEFAULT CHARACTER SET = utf8mb4 COLLATE = utf8mb4_0900_ai_ci, 
  MODIFY COLUMN `[field1]` char(3) NOT NULL, 
  MODIFY COLUMN `[field2]` char(3) NOT NULL;
```

On each of the ALTERs in the diff, it has `COLLATE = utf8mb4_0900_ai_ci`, and then on the tables that have not been changed, all of the char/varchar/text columns have a `MODIFY`, but there is no actual difference in those between .SQL and DB.

**Comment by evanelias** [comment]
That behavior is actually unrelated to the original bug here. Rather, it's just another side-effect of the situation mentioned in this logged warning: `Official arm64 Docker images for mysql:5.7 are not available. Substituting mysql:8.0 instead for workspace purposes, which may cause behavior differences.`

Your Skeema configuration uses `workspace=docker`, and you're running Skeema on an Apple Silicon (arm64 CPU) Mac. Using `workspace=docker` tells Skeema to use a Docker container on localhost for interpreting the meaning of your *.sql files. However, since there's no arm64 Docker image for MySQL 5.7 available, Skeema must use MySQL 8.0 instead. This causes behavior differences: the meaning of any given `CREATE TABLE` statement on MySQL 8 can be slightly different than on MySQL 5.7.

One major difference between the two versions is that the default collation for the utf8mb4 character set has changed. From my recent [blog post on schema and DDL differences between MySQL 5.7 and MySQL 8](https://www.skeema.io/blog/2024/02/28/mysql57-to-8-upgrade/):

> The utf8mb4 character set is the modern preferred choice for arbitrary textual data, since it can encode any unicode character. Previously, its default collation was utf8mb4_general_ci, which doesn’t conform to any UCA standard. This old collation considers all supplementary characters (such as emoji) to be equal in comparisons, which is especially problematic for uniqueness constraints.
> 
> MySQL 8 introduces new collations to solve this issue, and the default collation for utf8mb4 has been updated accordingly: it is now utf8mb4_0900_ai_ci, which is [based on UCA 9.0.0](https://dev.mysql.com/doc/refman/8.0/en/charset-unicode-sets.html). This is a positive improvement, but the default change causes upgrade complications with logical dumps, as well as schema management. If you execute a `CREATE TABLE` statement which includes `CHARACTER SET utf8mb4` columns, but doesn’t explicitly specify a `COLLATE` clause, that logical statement now differs in meaning between MySQL 5.7 vs MySQL 8.

(Most of the rest of that post is focused on *upgrading* from 5.7 to 8, so other parts of the post won't apply to your situation until you're actually upgrading your prod database.)

In terms of how to work around this problem, there are a few different solutions available to you:

a) Modify your .sql files so that they explicitly say `COLLATE utf8mb4_general_ci` after each time the `utf8mb4` character set is mentioned. This way, you aren't relying on the *default* collation, which differs between 5.7 and 8.0. However, the downside to this is `skeema pull` will overwrite this change each time.

b) Run `skeema` only from Intel/AMD machines for now, until your databases are upgraded to MySQL 8 in the future. Since AWS is massively increasing the billing rate for Aurora MySQL 5.7 in less than 9 months, many companies using Aurora 5.7 are planning to upgrade before that that time. (AWS already did this cost increase for regular non-Aurora RDS MySQL 5.7 a couple weeks ago.)

c) Switch to `workspace=temp-schema` instead of `workspace=docker`. This will run workspaces on the real target database. However, that may be disruptive to production workloads in MySQL 5.7 if you have a lot of tables and a large buffer pool, so proceed with extreme caution.

d) Switch to `workspace=temp-schema` combined with the `temp-schema-environment` premium option, to run workspaces on a real Aurora 5.7 cluster, just not your production one. This isn't available in Skeema Community Edition though.

**Comment by evanelias** [comment]
Skeema v1.11.2 is [now available](https://github.com/skeema/skeema/releases/tag/v1.11.2), and includes fixes for the panic issue as well as the unexpected utf8mb4_0900_ai_ci diffs.
